### PR TITLE
Problem: Race condition in IPC sockets

### PR DESCRIPTION
**Pull request by Kentzo**
Solution: Don't unlink file on close

File may not belong to the socket at that point.

Refs #2764

### CODE REVIEW

**Comment by Kentzo** [comment]
A number of tests failed under certain configurations. Should look for errors or these tests are known to fail?

**Comment by bluca** [comment]
Thanks, but could you please split the refactor and the behaviour changes in 2 different commits? This will make it easier to track eventual regressions.

**Comment by bluca** [comment]
Also the test failures are new and are those that use IPC, so yes please have a look, thanks

**Comment by Kentzo** [comment]
There are 3 changes:

1. ZMQ_HAVE_IPC
2. cleanup in ipc_listener
3. bufgix

I can add 2 commits with 1 and 2,3.

**Comment by Kentzo** [comment]
Strange that it only fails with autotools.

File: src/ipc_listener.cpp [hunk 1]
```diff
@@ -210,115 +211,130 @@ int zmq::ipc_listener_t::get_address (std::string &addr_)
 
 int zmq::ipc_listener_t::set_address (const char *addr_)
 {
-    //  Create addr on stack for auto-cleanup
+    printf("a\n");
+    if (options.use_fd != -1) {
```
Comment (bluca) [inline src/ipc_listener.cpp]: this was done later otherwise the address is not saved in the endpoint class variable

Comment (bluca) [reply src/ipc_listener.cpp]: ditto for filename

Comment (Kentzo) [reply src/ipc_listener.cpp]: If understood the code correctly, when use_fd is set address is irrelevant.

Comment (bluca) [reply src/ipc_listener.cpp]: it's used for the events, for example, like a few lines below

**Comment by Kentzo** [comment]
I undid all of the refactoring as unfortunately don't have time to mess with it.

However, there seems to be a bug when [reporting the close event](https://github.com/zeromq/libzmq/blob/01a3f3955c04803e0131ea5e7e08788a138924d8/src/ipc_listener.cpp#L320): by that time `s` will be [reassigned to `retired_fd`](https://github.com/zeromq/libzmq/blob/01a3f3955c04803e0131ea5e7e08788a138924d8/src/ipc_listener.cpp#L295).

**Comment by Kentzo** [comment]
@bluca Can I backport this fix to 4.1 and 4.2 (what repo)?

**Comment by Kentzo** [comment]
@bluca Do you think `test_security_zap` failed due to my changes?

**Comment by bluca** [comment]
> However, there seems to be a bug when reporting the close event: by that time s will be reassigned to retired_fd.

Good point, I can take care of that, thanks.

> @bluca Can I backport this fix to 4.1 and 4.2 (what repo)?

Sure, for 4.2 this is the right repo already

> @bluca Do you think test_security_zap failed due to my changes?

That's https://github.com/zeromq/libzmq/issues/2733 it's a known (tricky) problem

**Pull request merged**